File: Analysis/ScoreTrajectoryAnalyzer.swift
import Foundation

/// Generates insights from historical health score trajectory.
/// Uses stored daily analysis snapshots to detect long-term score trends.
struct ScoreTrajectoryAnalyzer {

    /// Generate insights from score history with optional category breakdown
    static func generateInsights(scoreHistory: [(date: Date, score: Int)], categoryScores: [HealthScore] = []) -> [Insight] {
        guard scoreHistory.count >= 7 else { return [] }
        var insights: [Insight] = []

        // 1. Overall trajectory over last 30 days
        if let trajectory = analyzeTrajectory(scoreHistory: scoreHistory, days: 30, categoryScores: categoryScores) {
            insights.append(trajectory)
        }

        // 2. Score momentum: is improvement accelerating or decelerating?
        if let momentum = analyzeMomentum(scoreHistory: scoreHistory) {
            insights.append(momentum)
        }

        // 3. Sustained high/low periods
        if let sustained = analyzeSustainedPeriod(scoreHistory: scoreHistory) {
            insights.append(sustained)
        }

        return insights
    }

    // MARK: - Trajectory (30-day score direction)

    private static func analyzeTrajectory(scoreHistory: [(date: Date, score: Int)], days: Int, categoryScores: [HealthScore] = []) -> Insight? {
        let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: Date()) ?? Date()
        let recent = scoreHistory.filter { $0.date >= cutoff }
        guard recent.count >= 5 else { return nil }

        // Compare first half vs second half (use dropFirst/dropLast to avoid skipping middle element)
        let mid = recent.count / 2
        let firstHalf = Array(recent.prefix(mid))
        let secondHalf = Array(recent.dropFirst(mid))

        let firstAvg = Double(firstHalf.map(\.score).reduce(0, +)) / Double(firstHalf.count)
        let secondAvg = Double(secondHalf.map(\.score).reduce(0, +)) / Double(secondHalf.count)

        guard firstAvg > 0 else { return nil }
        let changePercent = ((secondAvg - firstAvg) / firstAvg) * 100

        // Only report significant changes (>3%)
        guard abs(changePercent) > 3 else { return nil }

        let improving = changePercent > 0
        let absChange = String(format: "%.0f", abs(changePercent))
        let currentScore = Int(secondAvg.rounded())

        // Build category driver text for declining scores
        let categoryDriverText: String
        if !improving && !categoryScores.isEmpty {
            let sorted = categoryScores
                .filter { $0.score < 70 }
                .sorted { $0.score < $1.score }
                .prefix(3)
            if !sorted.isEmpty {
                let driverList = sorted.compactMap { score -> String? in
                    guard let cat = score.category else { return nil }
                    return "\(cat.displayName) (\(score.score)/100)"
                }.joined(separator: ", ")
                categoryDriverText = " Pulled down by: \(driverList)."
            } else {
                categoryDriverText = ""
            }
        } else {
            categoryDriverText = ""
        }

        return Insight(
            metric: .restingHeartRate, // Representative metric for overall health
            title: improving ? Copy.Analysis.ScoreTrajectory.healthScoreTrendingUp : Copy.Analysis.ScoreTrajectory.healthScoreDeclining,
            summary: "Your overall health score has \(improving ? "improved" : "declined") \(absChange)% over the last \(days) days. Current average: \(currentScore).\(categoryDriverText)",
            recommendation: improving
                ? "Health score averaged \(currentScore) over the last \(days) days, up \(absChange)% from earlier in the period.\(categoryDriverText)"
                : "Health score averaged \(currentScore) over the last \(days) days, down \(absChange)% from earlier in the period.\(categoryDriverText)",
            severity: improving ? .info : .warning,
            trend: improving ? .improving : .declining,
            currentValue: secondAvg,
            baselineValue: firstAvg,
            deviationPercent: changePercent,
            category: .scoreTrajectory
        )
    }

    // MARK: - Momentum (acceleration of change)

    private static func analyzeMomentum(scoreHistory: [(date: Date, score: Int)]) -> Insight? {
        // Need at least 3 weeks of data for acceleration
        guard scoreHistory.count >= 21 else { return nil }

        let recent = Array(scoreHistory.suffix(21))

        // Split into 3 weeks
        let week1 = Array(recent.prefix(7))
        let week2 = Array(recent.dropFirst(7).prefix(7))
        let week3 = Array(recent.suffix(7))

        let avg1 = Double(week1.map(\.score).reduce(0, +)) / 7.0
        let avg2 = Double(week2.map(\.score).reduce(0, +)) / 7.0
        let avg3 = Double(week3.map(\.score).reduce(0, +)) / 7.0

        let change1to2 = avg2 - avg1
        let change2to3 = avg3 - avg2
        let acceleration = change2to3 - change1to2

        // Only report if acceleration is significant (>2 points)
        guard abs(acceleration) > 2 else { return nil }

        let accelerating = acceleration > 0
        let isImproving = change2to3 > 0

        // Interesting case: improvements are accelerating or declines are accelerating
        if accelerating && isImproving {
            return Insight(
                metric: .restingHeartRate,
                title: Copy.Analysis.ScoreTrajectory.improvementAccelerating,
                summary: Copy.Analysis.ScoreTrajectory.gainsPickingUpSpeed,
                recommendation: "Week-over-week score change accelerated: \(String(format: "%.1f", change1to2)) pts to \(String(format: "%.1f", change2to3)) pts. Current average: \(String(format: "%.0f", avg3)).",
                severity: .info,
                trend: .improving,
                currentValue: avg3,
                baselineValue: avg1,
                deviationPercent: ((avg3 - avg1) / max(avg1, 1)) * 100,
                category: .scoreTrajectory
            )
        } else if !accelerating && !isImproving {
            return Insight(
                metric: .restingHeartRate,
                title: Copy.Analysis.ScoreTrajectory.declineAccelerating,
                summary: Copy.Analysis.ScoreTrajectory.droppingFaster,
                recommendation: "Weekly decline rate increased from \(String(format: "%.1f", abs(change1to2))) pts to \(String(format: "%.1f", abs(change2to3))) pts. Score dropped from \(String(format: "%.0f", avg1)) to \(String(format: "%.0f", avg3)) over 3 weeks.",
                severity: .warning,
                trend: .declining,
                currentValue: avg3,
                baselineValue: avg1,
                deviationPercent: ((avg3 - avg1) / max(avg1, 1)) * 100,
                category: .scoreTrajectory
            )
        }

        return nil
    }

    // MARK: - Sustained Periods

    private static func analyzeSustainedPeriod(scoreHistory: [(date: Date, score: Int)]) -> Insight? {
        guard scoreHistory.count >= 7 else { return nil }

        let recent = Array(scoreHistory.suffix(14))

        // Check for sustained high (>85) or low (<60) scores
        let highDays = recent.filter { $0.score >= 85 }.count
        let lowDays = recent.filter { $0.score < 60 }.count

        if highDays >= 10 {
            let avg = Double(recent.map(\.score).reduce(0, +)) / Double(recent.count)
            return Insight(
                metric: .restingHeartRate,
                title: Copy.Analysis.ScoreTrajectory.consistentlyStrongHealth,
                summary: "Your health score has been 85+ for \(highDays) of the last \(recent.count) days. This is excellent.",
                recommendation: "Score at 85+ for \(highDays) of the last \(recent.count) days, averaging \(String(format: "%.0f", avg)).",
                severity: .info,
                trend: .improving,
                currentValue: avg,
                baselineValue: 85,
                deviationPercent: ((avg - 85) / 85) * 100,
                category: .scoreTrajectory
            )
        }

        if lowDays >= 7 {
            let avg = Double(recent.map(\.score).reduce(0, +)) / Double(recent.count)
            return Insight(
                metric: .restingHeartRate,
                title: Copy.Analysis.ScoreTrajectory.extendedLowScorePeriod,
                summary: "Your health score has been below 60 for \(lowDays) of the last \(recent.count) days.",
                recommendation: "Score below 60 for \(lowDays) of the last \(recent.count) days, averaging \(String(format: "%.0f", avg)).",
                severity: .warning,
                trend: .declining,
                currentValue: avg,
                baselineValue: 60,
                deviationPercent: ((avg - 60) / 60) * 100,
                category: .scoreTrajectory
            )
        }

        return nil
    }
}

// MARK: - InsightAnalyzer Conformance

extension ScoreTrajectoryAnalyzer: InsightAnalyzer {
    static var analyzerID: String { "scoreTrajectory" }
    static var insightCategory: InsightCategory { .scoreTrajectory }

    static func generateInsights(context: AnalysisContext) -> [Insight] {
        generateInsights(scoreHistory: context.scoreHistory, categoryScores: context.categoryScores)
    }
}

File: android-app/app/src/main/java/com/lasohealth/android/core/analysis/analyzers/ScoreTrajectoryAnalyzer.kt
package com.lasohealth.android.core.analysis.analyzers

import com.lasohealth.android.core.analysis.AnalysisContext
import com.lasohealth.android.core.analysis.AnalysisInsight
import com.lasohealth.android.core.analysis.DailySample
import com.lasohealth.android.core.analysis.InsightAnalyzer
import com.lasohealth.android.core.analysis.InsightDirective
import com.lasohealth.android.core.analysis.InsightPriority
import com.lasohealth.android.core.model.HealthMetric
import java.util.UUID
import kotlin.math.abs
import kotlin.math.max

/**
 * Generates insights from the overall health score trajectory.
 *
 * Analyses:
 *   1. Score trajectory over last 30 days (first-half vs second-half comparison)
 *   2. Score momentum (is improvement accelerating or decelerating over 3 weeks)
 *   3. Sustained high/low periods (>=85 for 10+ days, or <60 for 7+ days)
 *
 * Uses the health score stored as a pseudo-metric in the time series.
 * Falls back to HRV-based approximation if dedicated score series is unavailable.
 *
 * Ported from iOS ScoreTrajectoryAnalyzer.swift.
 */
class ScoreTrajectoryAnalyzer : InsightAnalyzer {

    override val name: String = "ScoreTrajectoryAnalyzer"

    override suspend fun generateInsights(context: AnalysisContext): List<AnalysisInsight> {
        // Build a score history from time series data.
        // Use all available baselines to compute a simplified composite score per day.
        val scoreHistory = buildScoreHistory(context)
        if (scoreHistory.size < 7) return emptyList()

        val insights = mutableListOf<AnalysisInsight>()

        analyzeTrajectory(scoreHistory, 30)?.let { insights += it }
        analyzeMomentum(scoreHistory)?.let { insights += it }
        analyzeSustainedPeriod(scoreHistory)?.let { insights += it }

        return insights
    }

    // ----- Score History Construction -----

    /**
     * Build a daily score history from the available time series data.
     * For each day, compute a simple composite score from key metrics relative to baselines.
     */
    private fun buildScoreHistory(context: AnalysisContext): List<Pair<Long, Int>> {
        val keyMetrics = listOf(
            HealthMetric.HEART_RATE_VARIABILITY,
            HealthMetric.RESTING_HEART_RATE,
            HealthMetric.SLEEP_DURATION,
            HealthMetric.STEPS,
            HealthMetric.EXERCISE_MINUTES,
        )

        // Build day -> metric -> value map
        val dayValues = mutableMapOf<Long, MutableMap<HealthMetric, Double>>()
        for (metric in keyMetrics) {
            val samples = context.timeSeries[metric] ?: continue
            for (sample in samples) {
                val day = startOfDay(sample.date)
                dayValues.getOrPut(day) { mutableMapOf() }[metric] = sample.value
            }
        }

        // Score each day
        return dayValues.mapNotNull { (day, metrics) ->
            if (metrics.size < 2) return@mapNotNull null
            var score = 0.0
            var count = 0

            for ((metric, value) in metrics) {
                val baseline = context.baselines[metric] ?: continue
                if (baseline.standardDeviation <= 0) continue

                val z = (value - baseline.mean) / baseline.standardDeviation
                // Map z-score to 0-100 range where 0 = -2 sigma, 100 = +2 sigma
                val rawScore = ((z + 2) / 4.0).coerceIn(0.0, 1.0) * 100.0

                // Flip for lower-is-better metrics (e.g., resting HR)
                val metricScore = if (metric.higherIsBetter) rawScore else 100.0 - rawScore
                score += metricScore
                count++
            }

            if (count == 0) null else day to (score / count).toInt().coerceIn(0, 100)
        }.sortedBy { it.first }
    }

    // ----- 1. Trajectory (30-day first-half vs second-half) -----

    private fun analyzeTrajectory(scoreHistory: List<Pair<Long, Int>>, days: Int): AnalysisInsight? {
        val now = System.currentTimeMillis()
        val cutoff = now - days.toLong() * DAY_MS
        val recent = scoreHistory.filter { it.first >= cutoff }
        if (recent.size < 5) return null

        val mid = recent.size / 2
        val firstHalf = recent.take(mid)
        val secondHalf = recent.drop(mid)

        val firstAvg = firstHalf.map { it.second.toDouble() }.average()
        val secondAvg = secondHalf.map { it.second.toDouble() }.average()
        if (firstAvg <= 0) return null

        val changePercent = ((secondAvg - firstAvg) / firstAvg) * 100.0
        if (abs(changePercent) <= 3) return null

        val improving = changePercent > 0
        val absChange = String.format("%.0f", abs(changePercent))
        val currentScore = secondAvg.toInt()

        return AnalysisInsight(
            id = UUID.randomUUID().toString(),
            title = if (improving) "Health Score Trending Up" else "Health Score Declining",
            detail = "Your overall health score has ${if (improving) "improved" else "declined"} " +
                "${absChange}% over the last $days days. Current average: $currentScore.",
            metric = HealthMetric.RESTING_HEART_RATE,
            priority = if (improving) InsightPriority.INFORMATIONAL else InsightPriority.MEDIUM,
            directive = if (improving) InsightDirective.MAINTAIN else InsightDirective.INFORMATIONAL,
        )
    }

    // ----- 2. Momentum (3-week acceleration) -----

    private fun analyzeMomentum(scoreHistory: List<Pair<Long, Int>>): AnalysisInsight? {
        if (scoreHistory.size < 21) return null
        val recent = scoreHistory.takeLast(21)

        val week1 = recent.take(7).map { it.second.toDouble() }.average()
        val week2 = recent.drop(7).take(7).map { it.second.toDouble() }.average()
        val week3 = recent.takeLast(7).map { it.second.toDouble() }.average()

        val change1to2 = week2 - week1
        val change2to3 = week3 - week2
        val acceleration = change2to3 - change1to2

        if (abs(acceleration) <= 2) return null

        val accelerating = acceleration > 0
        val isImproving = change2to3 > 0

        return when {
            accelerating && isImproving -> AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "Improvement Accelerating",
                detail = "Your health score gains are picking up speed. Week-over-week score change " +
                    "accelerated: ${String.format("%.1f", change1to2)} pts to ${String.format("%.1f", change2to3)} pts. " +
                    "Current average: ${String.format("%.0f", week3)}.",
                metric = HealthMetric.RESTING_HEART_RATE,
                priority = InsightPriority.INFORMATIONAL,
                directive = InsightDirective.MAINTAIN,
            )
            !accelerating && !isImproving -> AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "Decline Accelerating",
                detail = "Your health score is dropping faster. Weekly decline rate increased from " +
                    "${String.format("%.1f", abs(change1to2))} pts to ${String.format("%.1f", abs(change2to3))} pts. " +
                    "Score dropped from ${String.format("%.0f", week1)} to ${String.format("%.0f", week3)} over 3 weeks.",
                metric = HealthMetric.RESTING_HEART_RATE,
                priority = InsightPriority.HIGH,
                directive = InsightDirective.INFORMATIONAL,
            )
            else -> null
        }
    }

    // ----- 3. Sustained High/Low Periods -----

    private fun analyzeSustainedPeriod(scoreHistory: List<Pair<Long, Int>>): AnalysisInsight? {
        if (scoreHistory.size < 7) return null
        val recent = scoreHistory.takeLast(14)

        val highDays = recent.count { it.second >= 85 }
        val lowDays = recent.count { it.second < 60 }

        if (highDays >= 10) {
            val avg = recent.map { it.second.toDouble() }.average()
            return AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "Consistently Strong Health",
                detail = "Your health score has been 85+ for $highDays of the last ${recent.size} days. " +
                    "This is excellent. Average: ${String.format("%.0f", avg)}.",
                metric = HealthMetric.RESTING_HEART_RATE,
                priority = InsightPriority.INFORMATIONAL,
                directive = InsightDirective.MAINTAIN,
            )
        }

        if (lowDays >= 7) {
            val avg = recent.map { it.second.toDouble() }.average()
            return AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "Extended Low Score Period",
                detail = "Your health score has been below 60 for $lowDays of the last ${recent.size} days. " +
                    "Average: ${String.format("%.0f", avg)}.",
                metric = HealthMetric.RESTING_HEART_RATE,
                priority = InsightPriority.HIGH,
                directive = InsightDirective.INFORMATIONAL,
            )
        }

        return null
    }

    // ----- Helpers -----

    private fun startOfDay(epochMs: Long): Long {
        val cal = java.util.Calendar.getInstance().apply { timeInMillis = epochMs }
        cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
        cal.set(java.util.Calendar.MINUTE, 0)
        cal.set(java.util.Calendar.SECOND, 0)
        cal.set(java.util.Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    companion object {
        private const val DAY_MS = 86_400_000L
    }
}
